launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java

```diff
@@ -271,6 +271,8 @@ Map<String, String> getEffectiveConfig() throws IOException {
       Properties p = loadPropertiesFile();
       p.stringPropertyNames().forEach(key ->
         effectiveConfig.computeIfAbsent(key, p::getProperty));
+      effectiveConfig.putIfAbsent(SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH,
+        SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH_VALUE);
```

Comment:
Non SparkLauncher triggered (e.g., the Spark deamons you mentioned) classes are not affected, right?

Reply:
And, same question, why only driver default extra class is specified here?

Reply:
Yes, the following `spark-daemon.sh start` are not affected, @viirya .

```
$ grep 'spark-daemon.sh start' *
start-history-server.sh:exec "${SPARK_HOME}/sbin"/spark-daemon.sh start $CLASS 1 "$@"
start-master.sh:"${SPARK_HOME}/sbin"/spark-daemon.sh start $CLASS 1 \
start-worker.sh:  "${SPARK_HOME}/sbin"/spark-daemon.sh start $CLASS $WORKER_NUM \
```

This is a command on the driver side. The executors are supposed to use `spark.executor.extraClassPath` and `spark.executor.defaultExtraClassPath`.
> And, same question, why only driver default extra class is specified here?

Reply:
Hmm, why we need to specially deal with `SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH` here? I think Spark configs should be dealt together above:

```
p.stringPropertyNames().forEach(key ->
         effectiveConfig.computeIfAbsent(key, p::getProperty));
```

Reply:
This is required because when it's loaded from the file. `p.stringPropertyNames()` doesn't have both `SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH`. So, `.forEach` is not invoked at all.
```
Properties p = loadPropertiesFile();
      p.stringPropertyNames().
```

Reply:
Let me recheck.

Reply:
Yes, here.

When `getEffectiveConfig` is used like this in `SparkSubmitCommandBuilder.java`, `config` is a Java Map.
So, `String defaultExtraClassPath = config.get(SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH);` become `null` if we don't have this.

```
    Map<String, String> config = getEffectiveConfig();
    boolean isClientMode = isClientMode(config);
    String extraClassPath = isClientMode ? config.get(SparkLauncher.DRIVER_EXTRA_CLASSPATH) : null;
    String defaultExtraClassPath = config.get(SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH);
    if (extraClassPath == null || extraClassPath.trim().isEmpty()) {
```

Reply:
I see. Although this is only used by `SparkSubmitCommandBuilder` but not `SparkClassCommandBuilder` so I wonder if we should directly put it in `SparkSubmitCommandBuilder`.

Reply:
Hmm, but I saw you only changed `SparkSubmitCommandBuilder` to explicitly pick it up? If `SparkLauncher.DRIVER_EXTRA_CLASSPATH` is not set at all, these other classes won't use `SparkLauncher.DRIVER_DEFAULT_EXTRA_CLASS_PATH` too.

Reply:
Let me double-check with them `InProcessLauncher` and `SparkLauncher` once more. Last time, I checked it was okay, but your concern rings a bell to me.